pkg/controller/podgc/gc_controller.go

```diff
@@ -138,13 +138,29 @@ func isPodTerminating(pod *v1.Pod) bool {
 	return false
 }
 
-// Test Shutdown 
+// Test Shutdown
 func (gcc *PodGCController) gcTerminating(pods []*v1.Pod) {
 	klog.Infof("Test Shutdown: entering PodGCController:gcTerminating")
 	terminatedPods := []*v1.Pod{}
 	for _, pod := range pods {
 		if isPodTerminating(pod) {
-			terminatedPods = append(terminatedPods, pod)
+			node, err := gcc.nodeLister.Get(string(pod.Spec.NodeName))
+			if err != nil {
+				klog.Errorf("failed to get node:%s", err)
+				continue
+			}
+
+			// Add this pod to terminatedPods list only if the following conditions are met:
+			// 1. Node is not ready.
+			// 2. Node has `out-of-service` taint with `NoExecute` effect.
+			// 3. Pod tolerated the `out-of-service` taint with `NoExecute` effect.
+			taintToMatch := &v1.Taint{
+				Key:    nodeutil.OutOfServiceTaint,
+				Effect: v1.TaintEffectNoExecute,
+			}
+			if !nodeutil.IsNodeReady(node) && hasMatchingTaint(taintToMatch, node) && isPodTolerate(taintToMatch, pod) {
```

Comment:
What if the node as `out-of-service` taint but the effect is NOT `NoExecute`

Reply:
hasMatchingTaint --> Returns true if the taint matches exactly the same for `key` as well as `effect`

Reply:
"NoExecute" means evicting pods. So I think we should find the matching effect as well.

Reply:
Yes. The functions `hasMatchingTaint` find the matching effect as well. 
So -- `hasMatchingTaint` returns true if the node as a taint with key `out-of-service` and effect as `NoExecute` -- for all other combination it returns false.

Do you see any action item still here?